Automattic/jetpack
# Jetpack: improve plugin disconnection flow

This plugin is based from #21048

#### Changes proposed in this Pull Request:
* Replaces the current `<JetpackDisconnectModal />` with the `<DisconnectDialog />` component from the JS connection package.
* Uses the `<DisconnectDialog />` to display information from the previous disconnection modal in an updated design.
* Introduces a new small React app on the plugins page to use the same `<DisconnectDialog />` modal to provide information before the Jetpack plugin is deactivated.
* Preloads information needed in the disconnection flow into the initial React state to eliminate load time for the disconnection flow.

#### Jetpack Product Discussion
pbNhbs-1cB-p2

#### Does this pull request change what data or activity we track or use?
This PR does not track any new data or activity on its own.

#### Testing instructions:
* Take a look at the test plan for #21048 to see how the updated disconnection flow works. This PR builds on that one, so it will focus on the specific outputs related to the Jetpack plugin.
* Check out this branch in your local Jetpack dev environment.
* Make sure you have another plugin that uses the Jetpack connection installed and active ( like Backup or Boost )
* It's also helpful if your test site has a few site stats/ visits to see some of the output in this test.

**Testing disconnection from the plugin dashboard**
* Navigate to the Jetpack plugin dashboard.
* Scroll down to the "Connections" section and click on the "Manage site Connection" link
![Screen Shot 2021-10-22 at 4 01 33 PM](https://user-images.githubusercontent.com/18016357/138526765-0a9cbb28-c9c1-45ff-a08c-126a195eeed7.png)

* You should be presented with the updated disconnection flow that is using the `<DisconnectDialog /> component.`
* On the first screen, you should see first, a list of plugins that are using the Jetpack connection. Secondly, you should see a "Jetpack Benefits" section, that shows some high-level numbers about the site. If you have 2 or fewer stats here, you will also see a set of links about using more Jetpack features ( this is the same information from the previous disconnection flow ).
![Screen Shot 2021-10-22 at 4 10 09 PM](https://user-images.githubusercontent.com/18016357/138526738-5921246b-8681-4d07-bd8f-567d90545995.png)
![Screen Shot 2021-10-22 at 4 10 24 PM](https://user-images.githubusercontent.com/18016357/138526757-96604baf-7b02-4e8d-8df8-8b6f01415537.png)

* After a successful disconnect, you will see the disconnect confirmation screen, which will provide the option to give some feedback. You can also exit the flow here by clicking on "No thank you"
![Screen Shot 2021-10-22 at 3 55 35 PM](https://user-images.githubusercontent.com/18016357/138526787-bb71635a-e013-4a06-a99f-b12867680633.png)

* Click on the "Help us improve" button to go to the survey screen.
![Screen Shot 2021-10-14 at 12 05 25 PM](https://user-images.githubusercontent.com/18016357/137379571-840434ea-daaa-48fd-b407-cbfb0b42e701.png)

* On the survey screen, the "Submit Feedback" button should be disabled until a response is selected. Selecting a response should show the response in an active/ selected state that looks like this:
![Screen Shot 2021-10-14 at 12 05 37 PM](https://user-images.githubusercontent.com/18016357/137379602-977d30e0-509b-49e5-90eb-0cd1adff87e9.png)

* Press "Submit Feedback" to submit the response and move to the last step. ( note, the disconnect flow can be exited here without providing feedback by clicking on "Skip for now" )
* The survey response is sent to a WPCOM endpoint. Observe your network panel traffic for a POST request to a `feedback-survey` endpoint. You should receive a 200 response from this endpoint.

* The final step of the flow is the "Thank You" screen. You can click "Back to my website," which will exit the disconnect flow and call the passed `onDisconnect` prop - if provided.
![Screen Shot 2021-11-20 at 11 25 49 AM](https://user-images.githubusercontent.com/18016357/142734288-44ebe9d2-cb27-4944-af2d-3b267f8f687b.png)

* Confirm that the analytics events listed in the test plan for #21048 are recorded by tracks.
* After disconnecting, you should see the Connect screen for the Jetpack plugin:
![Screen Shot 2021-10-22 at 4 11 29 PM](https://user-images.githubusercontent.com/18016357/138526806-6387674b-bdd0-4239-877e-764dc4973706.png)

**Testing disconnection from the plugins page**
* Be sure to reconnect Jetpack.
* Navigate to the plugins page for your test site at `wp-admin/plugins.php`
* After the JS for the page has loaded, the "Deactivate" link for the Jetpack plugin will change to say "Disconnect and Deactivate"
![Screen Shot 2021-10-22 at 4 12 44 PM](https://user-images.githubusercontent.com/18016357/138526845-a82efbd3-2141-4ad7-8f55-60aa03d45969.png)

* Clicking on this link will present you with the new Disconnection modal.
* The red button in the modal should read "Disconnect and Deactivate" in this context.
* You should see the same output as you did in the previous test ( other connected plugins and Jetpack benefits information )
* Click on "Disconnect and Deactivate". 
* After a moment, the plugins page should reload and the Jetpack plugin will be deactivated.
* Note that the remainder of the disconnection flow is bypassed in the context of the plugins page.
